docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1038 @@
+}
+
+// BootstrapProvider is the Schema for the BootstrapProviders API
+type BootstrapProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   BootstrapProviderSpec   `json:"spec,omitempty"`
+   Status BootstrapProviderStatus `json:"status,omitempty"`
+}
+
+// BootstrapProviderSpec defines the desired state of BootstrapProvider
+type BootstrapProviderSpec struct {
+   ProviderSpec `json:",inline"`
+}
+
+// BootstrapProviderStatus defines the observed state of BootstrapProvider
+type BootstrapProviderStatus struct {
+   ProviderStatus `json:",inline"`
+}
+
+// ControlPlaneProvider is the Schema for the ControlPlaneProviders API
+type ControlPlaneProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   ControlPlaneProviderSpec   `json:"spec,omitempty"`
+   Status ControlPlaneProviderStatus `json:"status,omitempty"`
+}
+
+// ControlPlaneProviderSpec defines the desired state of ControlPlaneProvider
+type ControlPlaneProviderSpec struct {
+   ProviderSpec `json:",inline"`
+}
+
+// ControlPlaneProviderStatus defines the observed state of ControlPlaneProvider
+type ControlPlaneProviderStatus struct {
+   ProviderStatus `json:",inline"`
+}
+
+// InfrastructureProvider is the Schema for the InfrastructureProviders API
+type InfrastructureProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   InfrastructureProviderSpec   `json:"spec,omitempty"`
+   Status InfrastructureProviderStatus `json:"status,omitempty"`
+}
+
+// InfrastructureProviderSpec defines the desired state of InfrastructureProvider
+type InfrastructureProviderSpec struct {
+   ProviderSpec `json:",inline"`
+}
+
+// InfrastructureProviderStatus defines the observed state of InfrastructureProvider
+type InfrastructureProviderStatus struct {
+   ProviderStatus `json:",inline"`
+}
+```
+
+Below you can find details about `ProviderSpec`,  `ProviderStatus`, which is
+shared among all the provider types - Core, Bootstrap, ControlPlane, and
+Infrastructure.
+
+```golang
+// ProviderSpec defines the desired state of the Provider.
+type ProviderSpec struct {
+   // Version indicates the provider version.
+   // +optional
+   Version string `json:"version,omitempty"`
+
+   // Manager defines the properties that can be enabled on the controller manager for the provider.
+   // +optional
+   Manager ManagerSpec `json:"manager,omitempty"`
+
+   // Deployment defines the properties that can be enabled on the deployment for the provider.
+   // +optional
+   Deployment DeploymentSpec `json:"deployment,omitempty"`
+
+   // SecretName is the name of the Secret providing the configuration
+   // variables for the current provider instance, like e.g. credentials.
+   // Such configurations will be used when creating or upgrading provider components.
+   // +optional
+   SecretName *string
```

Comment:
Let's try to get rid of secret or any credential management. The operator should defer to the providers for any credentials management, if we do find a way to make it global and interface-able we can do that in a future iteration